engine/src/main/java/io/zeebe/engine/processing/message/ProcessInstanceSubscriptionCorrelateProcessor.java

```diff
@@ -0,0 +1,158 @@
+          + "but it is already closing";
+
+  private final ProcessInstanceSubscriptionState subscriptionState;
+  private final SubscriptionCommandSender subscriptionCommandSender;
+  private final ProcessState processState;
+  private final ElementInstanceState elementInstanceState;
+  private final StateWriter stateWriter;
+  private final TypedRejectionWriter rejectionWriter;
+  private final EventHandle eventHandle;
+
+  public ProcessInstanceSubscriptionCorrelateProcessor(
+      final ProcessInstanceSubscriptionState subscriptionState,
+      final SubscriptionCommandSender subscriptionCommandSender,
+      final MutableZeebeState zeebeState,
+      final Writers writers) {
+    this.subscriptionState = subscriptionState;
+    this.subscriptionCommandSender = subscriptionCommandSender;
+    processState = zeebeState.getProcessState();
+    elementInstanceState = zeebeState.getElementInstanceState();
+    stateWriter = writers.state();
+    rejectionWriter = writers.rejection();
+
+    eventHandle =
+        new EventHandle(zeebeState.getKeyGenerator(), zeebeState.getEventScopeInstanceState());
+  }
+
+  @Override
+  public void processRecord(
+      final TypedRecord<ProcessInstanceSubscriptionRecord> command,
+      final TypedResponseWriter responseWriter,
+      final TypedStreamWriter streamWriter,
+      final Consumer<SideEffectProducer> sideEffect) {
+
+    final var subscriptionRecord = command.getValue();
+    final var elementInstanceKey = subscriptionRecord.getElementInstanceKey();
+
+    final ProcessInstanceSubscription subscription =
+        subscriptionState.getSubscription(
+            elementInstanceKey, subscriptionRecord.getMessageNameBuffer());
+
+    if (subscription == null) {
+      rejectCommand(command, RejectionType.NOT_FOUND, NO_SUBSCRIPTION_FOUND_MESSAGE);
+
+    } else if (subscription.isClosing()) {
+      rejectCommand(command, RejectionType.INVALID_STATE, ALREADY_CLOSING_MESSAGE);
+
+    } else {
+      final var elementInstance = elementInstanceState.getInstance(elementInstanceKey);
+      final var canTriggerElement = eventHandle.canTriggerElement(elementInstance);
+
+      if (!canTriggerElement) {
+        rejectCommand(command, RejectionType.INVALID_STATE, NO_EVENT_OCCURRED_MESSAGE);
+
+      } else {
+        // TODO (saig0): reuse the subscription record in the state (#6533)
+        subscriptionRecord
+            .setElementId(subscription.getTargetElementId())
+            .setInterrupting(subscription.shouldCloseOnCorrelate());
+
+        stateWriter.appendFollowUpEvent(
+            command.getKey(), ProcessInstanceSubscriptionIntent.CORRELATED, subscriptionRecord);
+
+        final var catchEvent =
+            getCatchEvent(elementInstance.getValue(), subscriptionRecord.getElementIdBuffer());
+        eventHandle.activateElement(
+            streamWriter, catchEvent, elementInstanceKey, elementInstance.getValue());
```

Comment:
I'm pretty sure this should be:
```suggestion
            stateWriter, catchEvent, elementInstanceKey, elementInstance.getValue());
```